docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,164 @@
+```
+
+## Versioning
+
+A mandatory `version` field in the YAML file expresses major and minor versions (e.g., 1.0, 1.1, 2.0).
+
+Compatibility rules
+
+1. If the major version of the YAML file is higher than the parser can handle, the parser should reject the file.
+2. If the minor version of the YAML file is higher than the highest version the parser is aware of, the parser should read the file but ignore any unrecognized content.
+3. If the file’s version is lower than the parser version, the parser should provide default values for any configuration introduced in later versions.
+
+Example parsing logic:
+
+```python
+def parse_custom_flow(content):
+    version = content['version']
+    major, minor = map(int, version.split('.'))
+
+    if major > PARSER_MAJOR:
+        raise IncompatibleVersionError("The custom flow file format is from a future major version.")
+    elif major <= PARSER_MAJOR and minor > PARSER_MINOR:
+        logger.warning("The custom flow file may have new features that will be ignored.")
+```
+
+### Flow Params
+
+The following parameters are supplied to a Flow constructor and used by the render() method:
+
+- client - an OpenAI completions API client for talking to the teacher model via the serving backend (i.e. llama-cpp or vLLM)
+- model_family - e.g. mixtral or merlinite
+- model_id - a path name for the specific teacher model being used
+- num_instructions_to_generate - how many samples to generate
+- batched - whether the model serving backend supports batched mode
+
+For now, we assume that the `Block` classes know how to use these parameters and there is no need to use any sort of templating in the custom flows that could use these parameters.
+
+### Model Prompts
+
+Based on whether model_family is mixtral or merlinite, a different prompt is used with the teacher model
+
+```python
+_MODEL_PROMPT_MIXTRAL = "<s> [INST] {prompt} [/INST]"
+_MODEL_PROMPT_MERLINITE = "'<|system|>\nYou are an AI language model developed by IBM Research. You are a cautious assistant. You carefully follow instructions. You are helpful and harmless and you follow ethical guidelines and promote positive behavior.\n<|user|>\n{prompt}\n<|assistant|>\n'"
+```
+
+For now, we assume that the `LLMBlock` class will choose the appropriate model prompt based on the family and that there is no need to specify a custom prompt.
+
+### Prompt Config Files
+
+Every LLMBlock references a separate prompt config file, and presumably a custom pipeline will provide custom prompt configs too.
+
+These prompt config files are quite simple YAML files - they contain a single object with system, introduction, principles, examples, and generation keys. See e.g. src/instructlab/sdg/configs/skills/freeform_questions.yaml
+
+We will continue to use these config files unchanged, and custom files can be specified with an absolute path. Relative paths are assumed to be relative to the Python package e.g. `configs/skills/...`.
+
+### Future - Model Serving
```

Comment:
I wrote more on this topic over here. I did it separately, but by the end I felt like maybe it's content to be integrated back into this one.

https://github.com/instructlab/dev-docs/pull/115

The good news is that I think what we need is already supported by `ilab`.

Reply:
Yeah, at the very least I need to capture here that this will work:

> ```
> version: "1.0"
> models:
>   - name: myfunkyadaptor
>     description: a funky adaptor for generating questions
> block_configs:
>   - block_type: LLMBlock
>     block_config:
>       block_name: gen_questions
>       config_path: configs/skills/freeform_questions.yaml
>       add_num_samples: True
>       gen_kwargs:
>         model_id: myfunkyadatpor
>       output_cols:
>         - question
>     drop_duplicates:
>       - question
> ```

And then the serving config just needs to have `--lora-modules=myfunkyadapter=path/to/my_custom_adapter`

Reply:
Ok, see commit 6457751 for how this impacts the pipeline config files

I started to incorporate #115 but it felt like the main point of that doc is the `config.yaml` impact, and it's better to describe that separately

Reply:
thanks the additions in that commit resolve this thread for me